[#12571] Instructors Edit Feedback Session: Instructor is able to edit submission opening time to an earlier timing

dlimyy opened the pull request:
Fixes #12571 

**Outline of Solution**

I created a new method called `triggerSubmissionOpeningDateModelChange` which is triggered when the instructor chooses a date in the date picker for the submission opening time. When the instructor changes the date to the earliest possible date, the `submissionStartTime` of the model will change accordingly to the earliest possible time for that date. Since the hours that the instructor are all in 0 minutes(with the exception of 23:59), I created another function `configureSubmissionOpeningTime` to configure the time properly.

I have also fixed the bug where the saved button is disabled when there is an error message and the fields are uneditable(This bug should not be able to be replicated here since I fixed the above bug but this fix is a safeguard against future bugs)

dlimyy commented:
I have fixed the bug regarding the case where the save button is disabled when the user passes in an invalid submission opening time. As seen below, the fields are now editable when there is an error.

https://github.com/TEAMMATES/teammates/assets/97420966/1bf5b137-614d-4eea-b231-d294d9be92f9

dlimyy commented:
This PR is also now ready for review

cedricongjh commented on src/web/app/components/session-edit-form/session-edit-form.component.ts:
```diff
@@ -134,6 +134,35 @@ export class SessionEditFormComponent {
+   * Configures the time for the submission opening time.
+   */
+  configureSubmissionOpeningTime(time : TimeFormat) : void {
+    if (time.hour < 24 && time.minute > 0) {
+      // Case where minutes is not 0 since the earliest time with 0 minutes is the hour before
+      time.hour += 1;
+      time.minute = 0;
+    } else if (time.hour === 23 && time.minute > 0) {
+      time.minute = 59;
+    }
```
time.hour === 23 && time.minute > 0 falls under the if branch case as well, I believe that the branches should be switched, could we also add unit tests for this method as well? (especially given that it deals with time)

cedricongjh commented on src/web/app/components/session-edit-form/session-edit-form.component.ts:
```diff
@@ -134,6 +134,35 @@ export class SessionEditFormComponent {
+   */
+  triggerSubmissionOpeningDateModelChange(field: string, date: DateFormat): void {
+    const minDate: DateFormat = this.minDateForSubmissionStart;
+    if (JSON.stringify(minDate) === JSON.stringify(date)) {
```
not a huge fan of this although it works, could we add a DateFormat comparision function in DateTimeService and use that here? (do add unit tests for the comparision function as well!)

cedricongjh reviewed (changes requested):
Thank you for your PR! (And great job finding this edge case)

Left a few comments that I think should be addressed.

Also, I've discovered that when switching to the min date, the time always defaults to the min time, even when the time selected is valid (e.g. 13th Sept, 2359 is selected, but switching to 12th Sept (the minDate) causes the time to change)

Do fix the issues and re-request a review!

dlimyy commented:
Hi @cedricongjh, thanks for reviewing of my PR and pointing out the bugs in my code! I have fixed the changes pointed out and added unit tests for all the methods requested. In addition, I have also solved the problem where switching the date to min date will always cause the time to revert to min time.

cedricongjh commented on src/web/app/components/session-edit-form/session-edit-form.component.ts:
```diff
@@ -134,6 +134,39 @@ export class SessionEditFormComponent {
+      this.model.submissionStartTime = minTime;
+    }
+
+    this.modelChange.emit({
+      ...this.model,
+      [field]: date,
```
nit: perhaps we could use `triggerModelChange` here

cedricongjh reviewed (changes requested):
Besides that one nit, LGTM!

dlimyy commented:
Hi @cedricongjh, I have changed it to triggerModelChange. Ready for review!

cedricongjh reviewed (approved):
LGTM, thank you for your contribution!

jasonqiu212 reviewed (approved):
LGTM! Thank you for spotting this edge case and fixing it :)
